libs/sdk-ui-dashboard/src/presentation/topBar/menuButton/DefaultMenuButton.tsx

```diff
@@ -26,24 +28,151 @@ const bubbleAlignPoints: IAlignPoint[] = [{ align: "cl tr" }];
 export const DefaultMenuButton = (props: IMenuButtonProps): JSX.Element | null => {
     const { menuItems } = props;
     const [isOpen, setIsOpen] = useState(false);
+    const [autofocusSubmenu, setAutofocusSubmenu] = useState(false);
     const intl = useIntl();
     const tooltipText = intl.formatMessage({ id: "controlButtons.options.tooltip" });
+    const backLabel = intl.formatMessage({ id: "controlButtons.options.back" });
+    const closeLabel = intl.formatMessage({ id: "controlButtons.options.close" });
+    const menuWrapperRef = useRef<HTMLDivElement>(null!);
+    const [parentItemId, setParentItemId] = useState<string | null>(null);
+    const menuItemRefs = useRef<Map<string, HTMLElement>>(new Map());
 
-    // generate unique anchor class name to open dropdown next to the correct button if app uses multiple
-    // dashboard components
-    const dropdownAnchorClassName = useMemo(() => `dash-header-options-anchor-${uuid()}`, []);
+    const setMenuItemRef = useCallback(
+        (itemId: string) => (element: HTMLDivElement | HTMLButtonElement | null) => {
+            if (element) {
+                menuItemRefs.current.set(itemId, element);
+            } else {
+                menuItemRefs.current.delete(itemId);
+            }
+        },
+        [],
+    );
+
+    // generate unique IDs for accessibility and dropdown positioning
+    const idPrefix = useId();
+    const buttonId = `${idPrefix}-menu-button`;
+    const menuId = `${idPrefix}-menu`;
+    const dropdownAnchorClassName = useMemo(
+        () => `dash-header-options-anchor-${idPrefix.slice(1)}`,
+        [idPrefix],
+    );
 
     const onMenuButtonClick = useCallback(() => {
         setIsOpen((prevIsOpen) => !prevIsOpen);
         setSelectedMenuItem(null);
     }, []);
 
+    const onClose = useCallback(() => {
+        setIsOpen(false);
+        setSelectedMenuItem(null);
+        setParentItemId(null);
+        setAutofocusSubmenu(false);
+    }, []);
+
+    const openSubmenu = useCallback((menuItem: IMenuButtonItemMenu) => {
+        setParentItemId(menuItem.itemId);
+        setAutofocusSubmenu(true);
+        setSelectedMenuItem(menuItem);
+    }, []);
+
+    const closeSubmenu = useCallback(() => {
+        //parentItemId is set to null later in effect focusing parent item
+        setAutofocusSubmenu(false);
+        setSelectedMenuItem(null);
+    }, []);
+
     const [selectedMenuItem, setSelectedMenuItem] = useState<IMenuButtonItemMenu | null>(null);
     const visibleMenuItems = useMemo(
         () => (selectedMenuItem?.items ?? menuItems).filter((item) => item.visible !== false),
         [menuItems, selectedMenuItem],
     );
 
+    const menuKeyboardNavigationHandler = makeMenuKeyboardNavigation<KeyboardEvent>({
+        onFocusFirst: () => {
+            const { firstElement } = getFocusableElements(menuWrapperRef.current ?? undefined);
```

Comment:
Why is there ` ?? undefined`. This means "return undefined if the previous part returned undefined", no?. I see it in other places as well.

Reply:
there is difference between null used as init value for ref and type of param of this function which is element?: Element. Thus this conversion. But I will treat it differently